### PR TITLE
RSDK-2014 - Refactor code

### PR DESCRIPTION
**Pull request by kkufieta**
Ticket: https://viam.atlassian.net/browse/RSDK-2014

Done:
* Refactor code:
    * Use [lidar package](https://github.com/viamrobotics/slam/blob/main/sensors/lidar/lidar.go)
    * Improve readability
    * Add tests

Tested through RDK.

### CODE REVIEW

File: viam-cartographer.go [hunk 1]
```diff
@@ -97,46 +93,108 @@ func init() {
...
-	cartoSvc *cartographerService,
-) error {
-	if !cartoSvc.useLiveData {
-		return nil
+	deps registry.Dependencies,
+	config config.Service,
+	logger golog.Logger,
+	bufferSLAMProcessLogs bool,
+	executableName string,
+	sensorValidationMaxTimeoutSec int,
```
Comment (tessavitabile) [inline viam-cartographer.go:104]: [opt] I preferred using functions to set these values, since I think this parameter list is getting too long.

File: viam-cartographer.go [hunk 2]
```diff
@@ -97,46 +93,108 @@ func init() {
...
+			if err := cartoSvc.Close(); err != nil {
+				logger.Errorw("error closing out after error", "error", err)
+			}
 		}
-		if !goutils.SelectContextOrWait(ctx, cameraValidationIntervalSec*time.Second) {
-			return ctx.Err()
+	}()
+
+	if cartoSvc.useLiveData {
+		if err := dim2d.ValidateGetAndSaveData(cancelCtx, cartoSvc.dataDirectory, lidar,
```
Comment (tessavitabile) [inline viam-cartographer.go:175]: What happens here if 'sensors' was empty? We haven't checked that sensors must be non-empty if useLiveData is true.

**Review by tessavitabile — CHANGES_REQUESTED** [review]
Just one major question

**Review by tessavitabile — APPROVED** [review]

File: helpers_test.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
+// Package viamcartographer_test tests the functions that require injected components (such as robot and camera)
```
[the remark this answers is not in the document]
Comment (kkufieta) [reply helpers_test.go:1]: That's a good idea, I will try/do.

File: viam-cartographer.go (hunk 2 above)
Comment (kkufieta) [reply viam-cartographer.go:175]: The evaluation of `useLiveData` ensures that the `sensors` array must contain a sensor when `use_live_data: true`: https://github.com/viamrobotics/slam/blob/519072a67f6a77476f0c32d2c524b42b75101ad1/config/config.go#L39-L41

**Review by kkufieta — COMMENTED** [review]
Great suggestions, thank you.

File: viam-cartographer.go (hunk 1 above)
Comment (zaporter-work) [reply viam-cartographer.go:104]: How do you feel about another struct being used as options? Maybe something like a `TimeoutConfig`? I agree that this list is getting too long, however the old solution had some non-pretty global state.

Comment (kkufieta) [reply viam-cartographer.go:104]: This became too complicated, provided I extracted much of the code and that it is used in multiple places. I was not confident that the old approach would not lead to unexpected behavior. I agree that the parameter list is too long, as well. Would creating a struct to collect those variables into one variable be a valid alternative?

File: viam-cartographer.go [hunk 4]
```diff
@@ -97,46 +93,108 @@ func init() {
...
+		port:                  port,
+		dataRateMs:            dataRateMsec,
+		mapRateSec:            mapRateSec,
+		cancelFunc:            cancelFunc,
+		logger:                logger,
+		bufferSLAMProcessLogs: bufferSLAMProcessLogs,
+		dev:                   svcConfig.Dev,
+	}
+
+	var success bool
```
Comment (zaporter-work) [inline viam-cartographer.go:165]:
```suggestion
	success := false
```

**Review by zaporter-work — APPROVED** [review]
Looking good! (though I think that helpers_test should be combined with the other test helpers)

File: helpers_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkufieta) [reply helpers_test.go:1]: This creates a circular dependency. Will abort.

[the remark this answers is not in the document]
Comment (zaporter-work) [reply helpers_test.go:1]: What circular dependency does this create?

File: internal/dim-2d/dim-2d.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	ctx context.Context,
+	sensors []string,
+	deps registry.Dependencies,
+	logger golog.Logger,
+) (lidar.Lidar, error) {
+	ctx, span := trace.StartSpan(ctx, "viamcartographer::internal::dim2d::getLidar")
+	defer span.End()
+
+	// An empty `sensors: []` array is allowed in offline mode.
+	if len(sensors) == 0 {
```
[the remark this answers is not in the document]
Comment (zaporter-work) [reply internal/dim-2d/dim-2d.go:54]: Ok. Resolving

File: helpers_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkufieta) [reply helpers_test.go:1]: The circular dependency got resolved by moving dim-2d_test.go to its own testing package! Doing it, after all :)

File: internal/dim-2d/dim-2d.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+		}
+		return "", err
+	}
+
+	filename := createTimestampFilename(dataDirectory, lidar.Name, ".pcd")
+	return filename, dataprocess.WritePCDToFile(pointcloud, filename)
+}
+
+// Creates a file for camera data with the specified sensor name and timestamp written into the filename.
+func createTimestampFilename(dataDirectory, primarySensorName, fileType string) string {
```
[the remark this answers is not in the document]
Comment (kkufieta) [reply internal/dim-2d/dim-2d.go:122]: I see. This function originally wrapped `dataprocess.CreateTimestampFilename` in various ways, depending on the slam service `subAlgo`. I think for cartographer it is unnecessary, though, so I removed it and pulled the contents into the `GetAndSaveData` function. Thanks for pointing this out!

File: internal/dim-2d/dim-2d.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+			return errors.Wrap(err, "error getting data from sensor")
+		}
+		if !goutils.SelectContextOrWait(ctx, time.Duration(sensorValidationIntervalSec)*time.Second) {
+			return ctx.Err()
+		}
+	}
+
+	for _, path := range paths {
+		if err := os.RemoveAll(path); err != nil {
+			return errors.Wrap(err, "error removing generated file during validation")
```
[the remark this answers is not in the document]
Comment (tessavitabile) [reply internal/dim-2d/dim-2d.go]: I think lidar validation would be clearer to a user, since they wouldn't be expected to look in the code at function names.

File: viam-cartographer_test.go [hunk 8]
```diff
@@ -1,392 +1,216 @@
...
 			Sensors:       []string{"invalid_sensor"},
 			ConfigParams:  map[string]string{"mode": "2d"},
-			DataDirectory: name,
-			DataRateMsec:  validDataRateMS,
+			DataDirectory: dataDir,
+			DataRateMsec:  dataRateMsec,
 			UseLiveData:   &_true,
 		}
 
-		_, err = createSLAMService(t, attrCfg, logger, false, false, testExecutableName)
```
Comment (EshaMaharishi) [inline viam-cartographer_test.go:294]: [q] Some tests like this one passed `success = false` to createSLAMService, but now CreateSLAMService always does what used to be `success = true`. I couldn't quickly tell why this change was ok - do you mind explaining?

**Review by EshaMaharishi — APPROVED** [review]
LGTM with questions that don't need to block merge if the answers are straightforward

File: viam-cartographer.go [hunk 9]
```diff
@@ -35,29 +32,28 @@ import (
...
 
+// Model is the model name of cartographer.
+var Model = resource.NewModel("viam", "slam", "cartographer")
+
 const (
-	// DefaultExecutableName contains the name of the cartographer grpc server executable.
-	DefaultExecutableName       = "carto_grpc_server"
-	defaultDataRateMsec         = 200
-	defaultMapRateSec           = 60
-	cameraValidationIntervalSec = 1.
```
[the remark this answers is not in the document]
Comment (kkufieta) [reply viam-cartographer.go:57]: Good catch! That was indeed a typo. I double-checked all values again and renamed the variables in the tests with a `Test...` prefix to distinguish them more from the default values and to make it clear that they're set to these values specifically for the tests.

File: viam-cartographer_test.go (hunk 8 above)
Comment (kkufieta) [reply viam-cartographer_test.go:294]: I'll gladly explain. Using `success` in the `createSLAMService` had no purpose. It simply ignored the value of `svc` in two test cases (for orbslam) when logically we expect the test to fail, but `success` is passed in as `true` (which makes no sense in itself). `svc` is `nil` in those two tests anyways - which is what I expect in my version of `createSLAMService`. 

Basically - when `err != nil` then `svc == nil` and vice versa. `success` simply ignored the value of `svc` in two test cases, when logically it didn't make sense to ignore the value of `svc`, cause we should expect it to be `nil`, and it is in fact `nil`.

Happy to discuss in person if it helps.

@zaporter-work this is also going to be relevant to you when you refactor orbslam. You can also safely remove `success` despite those two test cases.

File: viam-cartographer.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (EshaMaharishi) [reply viam-cartographer.go:57]: Sounds great! Very minor optional nit - we could name the variables xxxForTest, since TestXxx has some ambiguity with whether you're testing Xxx, or if Xxx is meant to be used in tests

File: viam-cartographer_test.go (hunk 8 above)
Comment (EshaMaharishi) [reply viam-cartographer_test.go:294]: Ah, I see! Thanks for explaining, and sounds good to me.
